pgrx/src/array.rs

```diff
@@ -332,15 +181,28 @@ impl RawArray {
         */
         unsafe {
             let ndim = self.ndim() as usize;
-            slice::from_raw_parts(ARR_DIMS(self.ptr.as_ptr()), ndim)
+            slice::from_raw_parts(port::ARR_DIMS(self.ptr.as_ptr()), ndim)
         }
     }
 
     /// The flattened length of the array over every single element.
     /// Includes all items, even the ones that might be null.
+    ///
+    /// # Panics
+    /// Panics if the Array's dimensions, multiplied together, exceed sizes Postgres can handle.
     #[inline]
     pub fn len(&self) -> usize {
-        self.len
+        // Calculating the product with i32 mirrors the Postgres implementation,
+        // except we can use checked_mul instead of trying to cast to 64 bits and
+        // hoping that doesn't also overflow on multiplication.
+        let dims = self.dims();
+        if dims.len() == 0 {
+            0
+        } else {
+            dims.into_iter()
+                .fold(Some(1i32), |prod, &d| prod.and_then(|m| m.checked_mul(d)))
+                .expect("Product of array dimensions should be <= i32::MAX") as usize
```

Comment:
This should probably be actually asserting on the size being less than `pg_sys::MaxArraySize`, since now that I look, that's smaller than `i32::MAX`... so tiresome.